app-libs/exchange-oracle/src/coingecko.rs

```diff
@@ -0,0 +1,176 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+use crate::{
+	error::{Error, Error::RestClientError},
+	ExchangeRateGet,
+};
+use itc_rest_client::{http_client::HttpClient, rest_client::RestClient, RestGet, RestPath};
+use log::*;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use url::Url;
+
+static COINGECKO_URL: &str = "https://api.coingecko.com";
+static COINGECKO_PARAM_CURRENCY: &str = "vs_currency";
+static COINGECKO_PARAM_COIN: &str = "ids";
+static COINGECKO_PATH: &str = "api/v3/coins/markets";
+
+/// REST client to make requests to CoinGecko.
+pub struct CoinGeckoClient {
+	client: RestClient<HttpClient>,
+}
+impl CoinGeckoClient {
+	fn new() -> Self {
+		let http_client = HttpClient::new(true, Some(Duration::from_secs(3u64)), None, None);
+		let url = Url::parse(COINGECKO_URL).unwrap(); //?
+		let rest_client = RestClient::new(http_client, url);
+		CoinGeckoClient { client: rest_client }
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CoinGeckoRoi {
+	times: Option<f64>,
+	currency: Option<String>,
+	percentage: Option<f64>,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CoinGeckoMarketStruct {
+	id: String,
+	symbol: String,
+	name: String,
+	image: Option<String>,
+	current_price: Option<f32>,
+	market_cap: Option<f64>,
+	market_cap_rank: Option<f32>,
+	fully_diluted_valuation: Option<f64>,
+	total_volume: Option<f64>,
+	high_24h: Option<f32>,
+	price_change_24h: Option<f32>,
+	price_change_percentage_24h: Option<f64>,
+	market_cap_change_24h: Option<f64>,
+	market_cap_change_percentage_24h: Option<f64>,
+	circulating_supply: Option<f64>,
+	total_supply: Option<f64>,
+	max_supply: Option<f64>,
+	ath: Option<f32>,
+	ath_change_percentage: Option<f32>,
+	ath_date: Option<String>,
+	atl: Option<f32>,
+	atl_change_percentage: Option<f32>,
+	atl_date: Option<String>,
+	roi: Option<CoinGeckoRoi>,
+	last_updated: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CoinGeckoMarket(pub Vec<CoinGeckoMarketStruct>);
+
+impl RestPath<String> for CoinGeckoMarket {
+	fn get_path(path: String) -> Result<String, itc_rest_client::error::Error> {
+		Ok(format!("{}", path))
+	}
+}
+
+impl ExchangeRateGet for CoinGeckoClient {
+	fn get_exchange_rate(
+		&mut self,
+		coin: String,
+		vs_currency: String,
```

Comment:
Make input parameters as generic as possible. That means here, try `&str` first. If you see that we have to use it as `String`, go with `&String`. Only use moved values (as you have currently), if you're taking ownership, otherwise use a borrow (`&`).